File: gfg/2024_8_august/c++/12_sum_of_middle_elements_of_two_sorted_arrays.cpp
// TIME COMPEXITY
// O(logn)

// SPACE COMPEXITY
// O(1)

// CODE

class Solution {
  public:
    int SumofMiddleElements(vector<int> &arr1, vector<int> &arr2) {
        int n = arr1.size();
        int st = 0;
        int end = n;
        
        while(st <= end){
            int m1 = (st+end) >> 1;
            int m2 = n - m1;
            
            int l1 = (m1==0) ? INT_MIN : arr1[m1-1];
            int l2 = (m2==0) ? INT_MIN : arr2[m2-1];
            int r1 = (m1 == n) ? INT_MAX : arr1[m1];
            int r2 = (m2 == n) ? INT_MAX : arr2[m2];
            
            if(l1 <= r2 && l2 <= r1) return max(l1,l2) + min(r1,r2);
            
            if(l1 > r2) end = m1 - 1;
            else st = m1+1;
        }
        
        return -1;
    }
};

File: gfg/2024_8_august/python/12_sum_of_middle_elements_of_two_sorted_arrays.py
# INTUITION
# The solution uses binary search to efficiently find the sum of the middle
# elements of two sorted arrays by identifying a partition between them. 
# By ensuring that elements on the total left side (including left of both 
# arrays) of the partition are less than or equal to those on the right, the
# algorithm calculates the middle elements as the maximum of the left-side 
# elements and the minimum of the right-side elements from both arrays. 

# This approach leverages the sorted nature of the arrays to find the correct 
# partition in logarithmic time, avoiding the need to merge the arrays directly.


# TIME COMPLEXITY
# O(log(n))


# SPACE COMPLEXITY
# O(1)


# CODE

class Solution:
    def sum_of_middle_elements(self, arr1, arr2):
        n = len(arr1)
        start = 0
        end = n
        
        while(start <= end):
            mid_of_arr1 = (start + end) >> 1
            mid_of_arr2 = n - mid_of_arr1
            
            left_of_arr1 = 0 if mid_of_arr1 == 0 else arr1[mid_of_arr1 - 1] 
            left_of_arr2 = 0 if mid_of_arr2 == 0 else arr2[mid_of_arr2 - 1] 
            right_of_arr1 = 10 ** 7 if mid_of_arr1 == n else arr1[mid_of_arr1] 
            right_of_arr2 = 10 ** 7 if mid_of_arr2 == n else arr2[mid_of_arr2]
            
            if ((left_of_arr1 <= right_of_arr2) and
                (left_of_arr2 <= right_of_arr1)):
                return (max(left_of_arr1, left_of_arr2) + 
                        min(right_of_arr1, right_of_arr2))
            
            if left_of_arr1 > right_of_arr2:
                end = mid_of_arr1 - 1
            else:
                start = mid_of_arr1 + 1
        
        return -1